java/ql/src/experimental/Security/CWE/CWE-094/SpringImplicitViewManipulation.ql

```diff
@@ -0,0 +1,40 @@
+/**
+ * @name Spring Implicit View Manipulation
+ * @description Untrusted input in a Spring View Controller can lead to RCE.
+ * @kind problem
+ * @problem.severity error
+ * @precision high
+ * @id java/spring-view-manipulation-implicit
+ * @tags security
+ *       external/cwe/cwe-094
+ */
+
+import java
+import SpringViewManipulationLib
+import DataFlow::PathGraph
+
+private predicate canResultInImplicitViewConversion(Method m) {
+  forex( |
+    m.getReturnType() instanceof TypeString or
+    m.getReturnType() instanceof VoidType
+  |
+    m.getParameter(_).(SpringRequestMappingParameter).isModelAttribute()
+  )
+  or
+  m.getReturnType() instanceof MapType
+  or
+  m.getReturnType().(RefType).hasQualifiedName("org.springframework.ui", "Model")
+  or
+  m.getReturnType() instanceof VoidType
+}
+
+from SpringRequestMappingMethod m
+where
+  thymeleafIsUsed() and
+  canResultInImplicitViewConversion(m) and
+  // A spring request mapping method which does not have response body annotation applied to it
+  forex( | m.getAnAnnotation().getType() instanceof SpringRequestMappingAnnotationType |
+    not m.getAnAnnotation().getType() instanceof SpringResponseBodyAnnotationType
+  ) and
+  not m.getParameterType(_) instanceof HttpServletResponse
```

Comment:
may it be the case that the servlet response is only used to set cookies or headers but the actual response content is delegated to a thymeleaf view?

Reply:
Per our slack discussion, when a request mapping method takes a `HttpServletResponse` field as a parameter, the response is assumed processed. Hence, the view manipulation vulnerabilities can't be exploited in this case.